plugins/catalog-backend/src/ingestion/HigherOrderOperations.ts

```diff
@@ -88,21 +85,33 @@ export class HigherOrderOperations implements HigherOrderOperation {
     // in the entities list. But we aren't sure what to do about those yet.
 
     // Write
-    if (!previousLocation) {
-      await this.locationsCatalog.addLocation(location);
-    }
-    if (readerOutput.entities.length === 0) {
-      return { location, entities: [] };
-    }
+    const entities = await this.database.transaction(async tx => {
+      if (!previousLocation) {
+        await this.locationsCatalog.addLocation(location, { tx });
+      }
+      if (readerOutput.entities.length === 0) {
+        return [];
+      }
 
-    const writtenEntities = await this.entitiesCatalog.batchAddOrUpdateEntities(
-      readerOutput.entities,
-      location.id,
-    );
+      const writtenEntities = await this.entitiesCatalog.batchAddOrUpdateEntities(
+        readerOutput.entities,
+        { locationId: location.id, tx },
```

Comment:
Still not too happy about exposing transactions at this layer. I think the main issue here is that we need to do the batch write and then read without modifying the DB? Wouldn't mind just adding any kind of `dryRunBatchAdd`-kinda method to the `EntitiesCatalog` for now. And then during a dry run we can easily just skip adding the location too right, so no need for transactions here.

We could also just make `batchAddOrUpdateEntities` a bit more powerful, giving it both a `dryRun` and a `output` options so that we don't even need the separate `entities` step below. That'd give us a place to solve both large entity volume issues and a method that can handle all of these use-cases and more.

Looking forward I think the `HigherOrderOperations` (or whatever we rename it to 😅 ) could also use its own `dryRun` method, something like

```ts
processEntities(opts: { entites: Entity[] }): EntityProcessingOutput
```

where `EntityProcessingOutput` (or w/e name) could eventually be a elaborate description on what effect processing the provided entities would have on the catalog, but right now it would just mirror the current `outputEntities` we're returning.

With that kind of method I think we'd be able to cover PR-feedback use-cases, as well as this dry-run use-case.

Dry-running a location `target` would essentially just be

```ts
const output = await catalog.processEntities({
  entities: [{
    apiVersion: 'backstage.io/v1alpha1',
    kind: 'Location',
    spec: {
      type: 'url'
      target,
    },
  }],
})
```

Thoughts on that @freben @benjdlambert, and think it's a sane direction @Fox32? 

Just to reiterate I'm thinking the best way forward here for now is to add both a `dryRun` option and an option to make `batchAddOrUpdateEntities` read back the entities it wrote.

Reply:
I'm not sure what kind of errors see would miss if we don't add the location, but I agree that skipping that part allows to remove the database from here. I will give it a try.